test/filewatch_test.erl

```diff
@@ -0,0 +1,76 @@
+-module(filewatch_test).
+-include_lib("eunit/include/eunit.hrl").
+
+-define(
+   rapid_events_test(Files, NEvents),
+
+   Dir = create_sandbox(?FUNCTION_NAME),
+   Pairs = [{Dir ++ File, File} || File <- Files],
+   {ok, P} = filewatch:start(self(), Pairs),
+   [start_sender(Path, NEvents) || {Path, _} <- Pairs],
+   State = collect_events(Files),
+   filewatch:stop(P),
+   ?assert(State)
+  ).
+
+one_file_few_rapid_events_test() ->
+    ?rapid_events_test(["testfile"], 5).
+
+one_file_many_rapid_events_test() ->
+    ?rapid_events_test(["testfile"], 1000).
+
+many_files_few_rapid_events_test() ->
+    ?rapid_events_test(["test1", "test2", "test3", "test4"], 5).
+
+many_files_many_rapid_events_test() ->
+    ?rapid_events_test(["test1", "test2", "test3", "test4"], 1000).
+
+%% Helper functions
+
+create_sandbox(Name) ->
+    Dir = ".tests/" ++ atom_to_list(Name) ++ "/",
+    ok = rm(Dir),
+    ok = filelib:ensure_dir(Dir),
+    Dir.
+
+rm(File) ->
+    case filelib:is_dir(File) of
+        true -> rmdir(File);
+        false ->
+            case filelib:is_file(File) of
+                true -> file:delete(File);
+                false -> ok
+            end
+    end.
+
+rmdir(Dir) ->
+    case file:list_dir(Dir) of
+        {error, enoent} -> ok;
+        {ok, Files} ->
+            lists:foreach(fun(F) -> rm(filename:join(Dir, F)) end, Files),
+            ok = file:del_dir(Dir)
+    end.
+
+start_sender(Path, N) ->
+    spawn(fun () -> send(Path, N) end).
+
+send(_Path, 0) -> ok;
+send(Path, N) ->
+    touch(Path),
```

Comment:
Did you verify this actually reproduces the same issues?  ISTR I did touch/rm alternation for a specific reason.